[ethash] remove mem::uninitialized

ordian opened the pull request:
Part of #10842. Supersedes #10853.

Closes #10842.

ordian commented on ethash/src/compute.rs:
```diff
@@ -233,8 +233,7 @@ fn hash_compute(light: &Light, full_size: usize, header_hash: &H256, nonce: u64)
 		},
-		// This is fully initialized before being read, see `let mut compress = ...` below
-		compress_bytes: unsafe { mem::uninitialized() },
+		compress_bytes: [0u8; MIX_WORDS],
```
```
❯ cargo bench --features=bench -- compute
bench_light_compute_memmap                                                                            
                        time:   [980.59 us 986.10 us 992.69 us]
                        change: [-3.7744% -1.4892% +0.5632%] (p = 0.20 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

bench_light_compute_memory                                                                             
                        time:   [1.0336 ms 1.0410 ms 1.0493 ms]
                        change: [-3.9993% -2.0714% +0.0881%] (p = 0.05 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
```

Vurich replied:
Amazing, must have been part of one of the LLVM upgrades since this code was written. I absolutely benchmarked it before and 0-initialising it was significantly slower. Wonder if there's anywhere else where we can remove uninit entirely now.

dvdplm commented on ethash/src/compute.rs:
```diff
@@ -282,9 +282,9 @@ fn hash_compute(light: &Light, full_size: usize, header_hash: &H256, nonce: u64)
-		// times and set each index, leaving the array fully initialized. THIS ONLY WORKS ON LITTLE-
-		// ENDIAN MACHINES. See a future PR to make this and the rest of the code work correctly on
+		// We iterate precisely `compress.len()` times and set each index,
+		// leaving the array fully initialized. THIS ONLY WORKS ON LITTLE-ENDIAN MACHINES.
```
Do you think it's possible to `if cfg(not(taget = "the-known-LE-targets")) { panic!("no no") }` here?

ordian replied:
add in 4cb0564

Vurich commented on ethash/src/compute.rs:
```diff
@@ -141,17 +141,22 @@ pub fn quick_get_difficulty(header_hash: &H256, nonce: u64, mix_hash: &H256, pro
+
+			// This is initialized in `keccak_256` below.
+			let mut hash = mem::MaybeUninit::<[u8; 32]>::uninit();
+			keccak_256::unchecked(hash.as_mut_ptr() as *mut u8, 32, buf.as_ptr(), buf.len());
```
Can you extract 32 to a constant so you don't have to ensure that the `32` in the definition of `hash` and the `32` in the `unchecked` call are the same? Before you could do `.len` but that's now impossible because of `MaybeUninit`.

Vurich commented on ethash/src/compute.rs:
```diff
@@ -277,12 +282,16 @@ fn hash_compute(light: &Light, full_size: usize, header_hash: &H256, nonce: u64)
 			unsafe { make_const_array!(MIX_WORDS / 4, &mut buf.compress_bytes) };
+		#[cfg(target_endian = "big")]
+		{
+			compile_error!("parity-ethereum currently only supports little-endian targets");
```
Great, this should have been done long ago. Re-reading the code, I think it might actually work on little-endian as long as you don't copy across the files we're reading from from little- to big-endian or vice-versa, although I can't be certain. Might be worth testing.

ordian commented on ethash/src/compute.rs:
```diff
@@ -152,9 +152,10 @@ pub fn quick_get_difficulty(header_hash: &H256, nonce: u64, mix_hash: &H256, pro
 
 			let buf = buf.assume_init();
 
+			const HASH_BYTES_LENGTH: usize = 32;
```
I'm bad at naming, so any other suggestions are welcome :)

Vurich replied:
It'll do

dvdplm replied:
Maybe `KECCAK_LEN`?

Vurich replied:
No, because `KECCAK_LEN` would be the total length right?

ordian commented:
Ok, so I've removed uninit completely, and running `cargo bench --features=bench -- compute` in ethash folder doesn't show any significant difference, but it would be nice if you could confirm this.

ordian commented on ethash/src/compute.rs:
```diff
@@ -135,23 +135,17 @@ pub fn quick_get_difficulty(header_hash: &H256, nonce: u64, mix_hash: &H256, pro
-			ptr::copy_nonoverlapping(&nonce as *const u64 as *const u8, buf[32..].as_mut_ptr(), 8);
+			let hash_len = header_hash.len();
+			buf[..hash_len].copy_from_slice(header_hash);
+			buf[hash_len..hash_len + mem::size_of::<u64>()].copy_from_slice(&nonce.to_ne_bytes());
```
I'm using here `to_ne_bytes` to preserve the previous behavior, but I guess it still only works for little-endian targets (don't have a machine to test this though).

dvdplm commented:
> would be nice if you could confirm this.

The benches on this branch:

```
     Running /Users/dvd/dev/parity/parity-ethereum/target/release/deps/basic-b975ddb423643e79
bench_light_compute_memmap
                        time:   [865.81 us 875.33 us 887.64 us]
                        change: [-7.0523% -2.5074% +1.7495%] (p = 0.30 > 0.05)
                        No change in performance detected.
Found 16 outliers among 100 measurements (16.00%)
  5 (5.00%) high mild
  11 (11.00%) high severe

bench_light_compute_memory
                        time:   [868.18 us 878.94 us 891.67 us]
                        change: [-4.1526% -1.4291% +1.4194%] (p = 0.33 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe
```

However, the same bench command on master is… weird. Here's the output:

```
     Running /Users/dvd/dev/parity/parity-ethereum/target/release/deps/basic-b975ddb423643e79
bench_light_compute_memmap
                        time:   [858.62 us 869.47 us 882.64 us]
                        change: [-4.8265% -1.4632% +1.5786%] (p = 0.41 > 0.05)
                        No change in performance detected.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe

bench_light_compute_memmap
                        time:   [858.14 us 867.39 us 878.12 us]
                        change: [-1.7316% +0.9385% +3.9310%] (p = 0.54 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  10 (10.00%) high mild
  3 (3.00%) high severe

Benchmarking bench_light_compute_memmap: Collecting 100 samples in estimated 2948.4 s (5050 iterations)^C
```

Notice the ~3000s estimate and the duplicate bench names.

ordian commented:
> the duplicate bench names

fixed in this PR in 5ef9344 

> Notice the ~3000s estimate 

yeah, I guess we can force criterion to do less iterations, added 7c2d729. At this point I'm starting to think about splitting the PR into bench and non-bench part, but it's small enough so far.
